File: day10_part2_2017.php

<?php

$input = file_get_contents("input.txt");

$lengths = [];
for ($i = 0; $i < strlen($input); $i++) {
    $lengths[] = ord($input[$i]);
}
$lengths = array_merge($lengths, [17, 31, 73, 47, 23]);

$list = range(0, 255);
$currentPosition = 0;
$skipSize = 0;

for ($round = 0; $round < 64; $round++) {
    foreach ($lengths as $length) {
        for ($i = 0; $i < $length / 2; $i++) {
            $start = ($currentPosition + $i) % 256;
            $end = ($currentPosition + $length - 1 - $i) % 256;
            [$list[$start], $list[$end]] = [$list[$end], $list[$start]];
        }
        $currentPosition = ($currentPosition + $length + $skipSize) % 256;
        $skipSize++;
    }
}

$denseHash = [];
for ($i = 0; $i < 256; $i += 16) {
    $xor = 0;
    for ($j = 0; $j < 16; $j++) {
        $xor ^= $list[$i + $j];
    }
    $denseHash[] = $xor;
}

$hexHash = bin2hex(call_user_func_array("pack", array_merge(["C*"], $denseHash)));

echo $hexHash . "\n";
?>

File: objectivec/day10_part2_2017.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSMutableArray<NSNumber *> *lengths = [NSMutableArray array];
        for (NSUInteger i = 0; i < input.length; i++) {
            [lengths addObject:@([input characterAtIndex:i])];
        }
        [lengths addObjectsFromArray:@[@17, @31, @73, @47, @23]];

        NSMutableArray<NSNumber *> *list = [NSMutableArray arrayWithCapacity:256];
        for (int i = 0; i < 256; i++) {
            [list addObject:@(i)];
        }

        NSUInteger currentPosition = 0, skipSize = 0;
        for (int round = 0; round < 64; round++) {
            for (NSNumber *length in lengths) {
                NSUInteger len = length.unsignedIntegerValue;
                for (NSUInteger i = 0; i < len / 2; i++) {
                    NSUInteger start = (currentPosition + i) % 256;
                    NSUInteger end = (currentPosition + len - 1 - i) % 256;
                    NSNumber *temp = list[start];
                    list[start] = list[end];
                    list[end] = temp;
                }
                currentPosition = (currentPosition + len + skipSize) % 256;
                skipSize++;
            }
        }

        NSMutableArray<NSNumber *> *denseHash = [NSMutableArray array];
        for (NSUInteger i = 0; i < 256; i += 16) {
            NSUInteger xor = 0;
            for (NSUInteger j = 0; j < 16; j++) {
                xor ^= list[i + j].unsignedIntegerValue;
            }
            [denseHash addObject:@(xor)];
        }

        NSMutableString *hexHash = [NSMutableString string];
        for (NSNumber *num in denseHash) {
            [hexHash appendFormat:@"%02x", num.unsignedIntegerValue];
        }

        NSLog(@"%@", hexHash);
    }
    return 0;
}